pkg/kubelet/lifecycle/handlers.go

```diff
@@ -117,7 +121,19 @@ func (hr *handlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
 			return "", fmt.Errorf("failed to find networking container: %v", status)
 		}
 		host = status.IPs[0]
+		podIP = host
 	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.ConsistentHTTPGetHandlers) {
+		req, err := httpprobe.NewRequestForHTTPGetAction(handler.HTTPGet, container, podIP, "lifecycle")
+		if err != nil {
+			return "", err
+		}
+		resp, err := hr.httpDoer.Do(req)
```

Comment:
I've been thinking about how to roll this change out safely in situations like the following:
* the cluster admin deciding whether to enable/disable the feature gate doesn't have visibility or control to all pod specs to sweep them for valid lifecycle configurations
* lifecycle configurations are incorrectly specifying `scheme: "https"` when speaking to http handler endpoints and working successfully, and the author is unaware of the error

For lifecycle handlers, I think we should:
1. fall back to `http` if an `https` request encounters an error that indicates it is speaking to an `http` endpoint (which the go client emits a [distinct error for](https://github.com/golang/go/blob/1f068f0dc7bc997446a7aac44cfc70746ad918e0/src/net/http/client.go#L263)), either by calling `hr.httpDoer.Do(req)` again with a request modified to talk to `http` or by continuing to the deprecated code path below.
2. when we successfully fall back to http from https:
    * log
    * increment a metric (cc @logicalhan for pointers on setting this up and naming it)
    * maybe emit an event

This will allow incorrectly configured lifecycle handlers to be detected and fixed gradually and nondisruptively

Since we're using an unauthenticated client that doesn't verify TLS, I don't think the security attributes of a fallback are that different. We could consider clearing any Authorization header when falling back from https → http if we wanted to be sure we weren't leaking tokens/basicauth credentials (which hopefully no one is putting in their pod specs, but 🤷‍♂️). Since headers weren't honored before, dropping the Authorization header when falling back to http wouldn't break anyone currently working.

Reply:
What should I do here? Do you want to hold this PR up until we hear back from @logicalchan?

I don't mind making the changes you suggest, but I will admit to be hesitant about even more changes to the behaviors here; I'm worried that I'll do the work and be left hanging again trying to get reviews and signoffs all the while the real bugs that this PR currently fixes remain in effect.

Reply:
go ahead and set up the retry/fallback with a TODO placeholder for any metric/event reaction.

Unfortunately there's a rich tradition of testing error strings from the stdlib http client :-/

I'm mobile now so I don't have references handy, but there are places where we don't have better options. We usually pair that with a unit test that exercises the condition so we notice if the stdlib changes messages in ways that break us.

Reply:
> I'm worried that I'll do the work and be left hanging again trying to get reviews and signoffs all the while the real bugs that this PR currently fixes remain in effect.

I'm sympathetic to that worry. If a synchronous session to knock out this bit would be helpful, I'd be happy to block off time for it, just let me know.

Reply:
> We usually pair that with a unit test that exercises the condition so we notice if the stdlib changes messages in ways that break us.

A pushed a unit test, but then I realized that while that verifies the behavior we want to check, it won't verify that the triggering error message remains the same across Go releases. The fix is probably easy enough, though. Is it acceptable  to use a real HTTP server that listens on a dynamic port in the unit tests or does this need to be done as an e2e test?

Reply:
I pushed a change to implement the unit test using a server that listens on a dynamic port. If that goes against the guidelines, then let me know and I'll work up an e2e test.